domain/application/service/resource.go

```diff
@@ -83,14 +97,91 @@ type ContainerImageMetadataState interface {
 	) (application.ContainerImageMetadata, error)
 }
 
+// AddResources adds specified resources to an application. It validates the
+// application ID and each resource's origin. Returns a result containing
+// added resources' UUIDs or an error if validation fails.
+//
+// The following error types can be expected to be returned:
+//   - [applicationerrors.ApplicationIDNotValid] is returned if the application
+//     ID is not valid.
+//   - [applicationerrors.CharmOriginNotValid] is returned if the charm origin
+//     ID is not valid.
+//   - [applicationerrors.ResourceNameNotValid] if no resource name is provided in
+//     the args or if it doesn't match with a charm declared resource.
+//   - [applicationerrors.ResourceNotFound] if no resource with name exists for
+//     given application.
+func (s *Service) AddResources(
+	ctx context.Context,
+	args resource.AddResourcesArgs,
+) (resource.AddResourcesResult, error) {
+	var validationErrors []error
+	if err := args.ApplicationID.Validate(); err != nil {
+		validationErrors = append(validationErrors, applicationerrors.ApplicationIDNotValid)
+	}
+	for i, res := range args.Resources {
+		if err := res.Origin.Validate(); err != nil {
+			validationErrors = append(validationErrors,
+				fmt.Errorf("resource %d: %w", i,
+					applicationerrors.CharmOriginNotValid))
+		}
+		if err := res.State.Validate(); err != nil {
+			validationErrors = append(validationErrors,
+				fmt.Errorf("resource %d: %w", i,
+					applicationerrors.ResourceStateNotValid))
+		}
+		if res.Name == "" {
+			validationErrors = append(validationErrors, fmt.Errorf("resource %d: %w", i,
+				applicationerrors.ResourceNameNotValid))
+		}
+	}
+	if err := errors.Join(validationErrors...); err != nil {
+		return resource.AddResourcesResult{}, err
+	}
+	return s.st.AddResources(ctx, args)
+}
+
+// DeleteApplicationResources removes the resources of a specified application.
+// It should be called after all resources have been unlinked from potential
+// units by DeleteUnitResources and their associated data removed from store.
+//
+// The following error types can be expected to be returned:
+//   - [applicationerrors.ApplicationIDNotValid] is returned if the application ID is not valid.
+//   - [applicationerrors.InvalidApplicationState] is returned is there is
```

Comment:
todo: find or create a new error. InvalidApplicationState appears to be specific to platforms, though it is not very descriptive.